### PR TITLE
Typescript conversion and Nextjs version upgrade

### LINKED ISSUE #57: Use TypeScript

### PR DESCRIPTION
**Pull request by samepant**
~Opening this early to get some eyes on the preliminary setup. Will be converting all the pages and components over the next few days.~

fixes #57

Ready for review 🥂

### CODE REVIEW

**Comment by samepant** [comment]
I've tested the components and navigation but havent been able to send a TX successfully (seeing this https://github.com/cosmos/cosmos-multisig-ui/issues/60) but it appears to be unrelated to the typescript conversion.

**Comment by samepant** [comment]
fixes #57

File: components/chainSelect/ChainSelect.tsx [hunk 1]
```diff
@@ -4,18 +4,40 @@ import { StargateClient } from "@cosmjs/stargate";
...
+
 const ChainSelect = () => {
   const { state, dispatch } = useAppContext();
 
   // UI State
   const [chainArray, setChainArray] = useState();
-  const [chainOptions, setChainOptions] = useState();
-  const [chainError, setChainError] = useState();
+  const [chainOptions, setChainOptions] = useState<ChainOption[]>([]);
+  const [chainError, setChainError] = useState<string | null>();
```
Comment (webmaster128) [inline components/chainSelect/ChainSelect.tsx:40]: Does this compile? What's the default value in this case?

File: components/chainSelect/ChainSelect.tsx [hunk 2]
```diff
@@ -166,17 +188,17 @@ const ChainSelect = () => {
     return secureNodes[0].address;
   };
 
-  const onChainSelect = (option) => {
+  const onChainSelect = (option: ChainOption) => {
     const index = chainOptions.findIndex((opt) => opt.label === option.label);
     setSelectValue(chainOptions[index]);
-    getChainInfo(chainArray[option.value]);
+    getChainInfo(chainArray![option.value]);
```
Comment (webmaster128) [inline components/chainSelect/ChainSelect.tsx:194]: Can we default `chainArray` to `{}`? Then we probably do not need this `!`.

File: components/dataViews/CompletedTransaction.tsx [hunk 3]
```diff
@@ -6,7 +6,7 @@ import Button from "../inputs/Button";
 import { explorerLinkTx } from "../../lib/displayHelpers";
 import { useAppContext } from "../../context/AppContext";
 
-const CompletedTransaction = ({ transactionHash }) => {
+const CompletedTransaction = ({ transactionHash }: { transactionHash: string }) => {
```
Comment (webmaster128) [inline components/dataViews/CompletedTransaction.tsx:9]: Can we create a `Props` interface here to make this code more readable?

File: components/dataViews/HashView.tsx [hunk 4]
```diff
@@ -3,10 +3,10 @@ import React from "react";
 import { abbreviateLongString } from "../../lib/displayHelpers";
 import CopyAndPaste from "./CopyAndPaste";
 
-const HashView = ({ hash, abbreviate }) => (
+const HashView = (props: { hash: string; abbreviate?: boolean }) => (
```
Comment (webmaster128) [inline components/dataViews/HashView.tsx:6]: A `Props` interface would be nice

File: components/dataViews/MultisigHoldings.tsx [hunk 5]
```diff
@@ -1,16 +1,25 @@
...
   return (
     <StackableContainer lessPadding fullHeight>
       <h2>Holdings</h2>
       <StackableContainer lessPadding lessMargin>
-        <span>{printableCoin(props.holdings, state.chain)}</span>
+        {props.holdings ? (
+          <span>{printableCoin(props.holdings, state.chain)}</span>
+        ) : (
+          <span>None</span>
+        )}
```
Comment (webmaster128) [inline components/dataViews/MultisigHoldings.tsx:22]: Nice!

File: components/dataViews/MultisigMembers.tsx [hunk 6]
```diff
@@ -3,7 +3,12 @@ import React from "react";
 import HashView from "./HashView";
 import StackableContainer from "../layout/StackableContainer";
 
-const MultisigMembers = (props) => (
+interface Props {
+  members: string[];
```
Comment (webmaster128) [inline components/dataViews/MultisigMembers.tsx:7]:
```suggestion
  /** Addresses of the multisig members */
  members: string[];
```

File: components/dataViews/ThresholdInfo.tsx [hunk 7]
```diff
@@ -1,8 +1,15 @@
...
 import StackableContainer from "../layout/StackableContainer";
 
-const ThresholdInfo = ({ signatures, account }) => (
+const ThresholdInfo = ({
+  signatures,
+  account,
+}: {
+  signatures: DbSavedSignature[];
+  account: Account;
+}) => (
```
Comment (webmaster128) [inline components/dataViews/ThresholdInfo.tsx:12]: `Props` interface would be nice

File: tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,20 @@
+{
+  "compilerOptions": {
+    "target": "es5",
```
Comment (webmaster128) [inline tsconfig.json:3]: This is what we use in CosmJS:

```suggestion
    "target": "es2018",
```

**Review by webmaster128 — COMMENTED** [review]
Very nice  🎉

I did not test it but the code looks solid.

File: components/chainSelect/ChainSelect.tsx (hunk 1 above)
Comment (samepant) [reply components/chainSelect/ChainSelect.tsx:40]: It does compile, default value is an empty array which leads to the Select UI to be empty until chain options are loaded from the registry.

**Comment by samepant** [comment]
@webmaster128 I added mods for all your comments, take a look and I'll squash this branch and merge. Thanks!!

File: components/chainSelect/ChainSelect.tsx (hunk 1 above)
Comment (webmaster128) [reply components/chainSelect/ChainSelect.tsx:40]: I mean this line with no default set:

```ts
  const [chainError, setChainError] = useState<string | null>();
```

Turns out the type becomes `string | null | undefined` and defaults to `undefined`.

I think there are two nice ways to fix:

1. Use `null`:   `const [chainError, setChainError] = useState<string | null>(null);`
2. Use `undefined`: `const [chainError, setChainError] = useState<string>();`

where 1. is probably what you wanted.

**Review by webmaster128 — APPROVED** [review]

File: components/chainSelect/ChainSelect.tsx (hunk 1 above)
Comment (samepant) [reply components/chainSelect/ChainSelect.tsx:40]: ah right, it should be null.

**Pull request merged**